Add react spinner to ENSURE_PAGE_SAFE

mirzal opened the pull request:
Some pages (Subscriptions and Red Hat Repositories for certain, maybe some others as well) use new react-based spinner widget. Our ensure_page_safe function did not take it into account and tried to proceed with automation even if page was not ready yet.

Function looks for element with class `loading-state` inside div with id `reactRoot`. That element is on page only when content is being loaded, i.e. it is removed from source on finish (instead of merely making it invisible). `reactRoot` is not present on other pages at all, so function will always return true on them.

oshtaier commented on airgun/browser.py:
```diff
@@ -513,11 +513,16 @@ class AirgunBrowserPlugin(DefaultPlugin):
          return (spinner === null) ? true : spinner.style["display"] == "none"
         }
+        function reactLoadingInvisible() {
+         react = document.querySelector("#reactRoot .loading-state")
```
Is whitespace really present before dot?

mirzal replied:
Yes. I'm looking for element with class that is child of `#reactRoot`.

This is HTML when data is being loaded:

```html
<div id="reactRoot">
  <div>
    <div class="container-fluid">
      <div class="row">
        <div class="col-sm-12">
          <h1>Red Hat Subscriptions</h1>
          <div class="toolbar-pf table-view-pf-toolbar-external row">
            <div class="col-sm-12">
              <form class="toolbar-pf-actions">
                <div class="toolbar-pf-filter form-group">
                  <div><label class="control-label sr-only">Search</label>
                    <div><span class="input-group"><input role="combobox" aria-autocomplete="list" aria-expanded="false" autocomplete="off" type="text" id="downshift-0-input" class="form-control" value=""><span class="input-group-btn"><button type="button" class="btn btn-default">Search</button></span></span>
                    </div>
                  </div>
                </div>
                <div class="toolbar-pf-action-right">
                  <div class="form-group">
                    <div class="tooltip-button-helper"><a href="subscriptions/add" class="btn btn-primary disabled" tabindex="-1" style="pointer-events: none;">Add Subscriptions</a></div><button type="button" class="btn btn-default">Manage Manifest</button><button type="button" class="btn btn-default">Export CSV</button>
                    <div
                      class="tooltip-button-helper"><button disabled="" type="button" class="btn btn-danger">Delete</button></div>
                </div>
            </div>
            </form>
          </div>
        </div>
        <div id="subscriptions-table" class="modal-container">
          <div class="loading-state">
            <div class="spinner spinner-lg"></div>
            <p>Loading</p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
</div>

</div>



<script defer="defer">
  //<![CDATA[
  $(tfm.reactMounter.mount('katello', '#reactRoot', []));
  //]]>
</script>
```

And after loading is finished:

```html
<div id="subscriptions-table" class="modal-container">
  <div>
    <table class="table table-striped table-bordered table-hover pf-table-inline-edit table-fixed">
      <thead>
        <tr class="">
          <th class="table-view-pf-select" aria-label="Select all rows"><label for="selectAll" class="control-label sr-only">Select all rows</label><input type="checkbox" id="selectAll"></th>
          <th class="">Name</th>
          <th class="">SKU</th>
          <th class="">Contract</th>
          <th class="">Start Date</th>
          <th class="">End Date</th>
          <th class="">Requires Virt-Who</th>
          <th class="">Consumed</th>
          <th class="">Entitlements</th>
        </tr>
      </thead>
      <tbody>
        <tr class="">
          <td class="table-view-pf-select"><label for="select0" class="control-label sr-only">Select row</label><input type="checkbox" id="select0"></td>
          <td><a href="/subscriptions/196/">Employee SKU</a></td>
          <td class="">ES0113909</td>
          <td class="">11341622</td>
          <td class="">2017-03-15 05:00:00 +0100</td>
          <td class="">2022-01-01 05:59:59 +0100</td>
          <td><span aria-hidden="true" class="fa fa-check"></span></td>
          <td class="">0</td>
          <td class="editable">
            <div class="input" role="textbox" tabindex="0">15</div>
          </td>
        </tr>
      </tbody>
    </table>
    <form class="content-view-pf-pagination table-view-pf-pagination clearfix">
      <div class="form-group">
        <div class="dropup pagination-pf-pagesize btn-group"><button id="pagination-row-dropdown" role="button" aria-haspopup="true" aria-expanded="false" type="button" class="dropdown-toggle btn btn-default">20 <span class="caret"></span></button>
          <ul role="menu" class="dropdown-menu" aria-labelledby="pagination-row-dropdown">
            <li role="presentation" class=""><a role="menuitem" tabindex="-1" href="#">5</a></li>
            <li role="presentation" class=""><a role="menuitem" tabindex="-1" href="#">10</a></li>
            <li role="presentation" class=""><a role="menuitem" tabindex="-1" href="#">15</a></li>
            <li role="presentation" class=""><a role="menuitem" tabindex="-1" href="#">25</a></li>
            <li role="presentation" class=""><a role="menuitem" tabindex="-1" href="#">50</a></li>
          </ul>
        </div><span>per page</span></div>
      <div class="form-group"><span><span class="pagination-pf-items-current">1-1</span>&nbsp;of&nbsp;<span class="pagination-pf-items-total">1</span></span>
        <ul class="pagination pagination-pf-back">
          <li class="disabled"><a href="#" title="First Page"><span aria-hidden="true" class="fa fa-angle-double-left i"></span></a></li>
          <li class="disabled"><a href="#" title="Previous Page"><span aria-hidden="true" class="fa fa-angle-left i"></span></a></li>
        </ul><label class="sr-only control-label">Current Page</label><input type="text" class="pagination-pf-page form-control" value="1"><span>&nbsp;of&nbsp;<span class="pagination-pf-pages">1</span></span>
        <ul class="pagination pagination-pf-forward">
          <li class="disabled"><a href="#" title="Next Page"><span aria-hidden="true" class="fa fa-angle-right i"></span></a></li>
          <li class="disabled"><a href="#" title="Last Page"><span aria-hidden="true" class="fa fa-angle-double-right i"></span></a></li>
        </ul>
      </div>
    </form>
  </div>
</div>
```

quarckster replied:
@mirzal @oshtaier RHAI pages also have the same or similar spinner, at least part `<div class="spinner spinner-lg"></div>` is the same #187

oshtaier commented:
that change is too generic, so will require some testing to not introduce the regression
Probably, we will need to run tests from each module and etc. I completely understand that can be redundant, but it is not.

oshtaier commented:
sure, you can sleep meanwhile :)

We had a job before on a Jenkins for 6.3, not sure what's about 6.4 to run scenario like that

Anyway, I appreciate if you can spend some time on things you mentioned in the first paragraph, and we can talk about tests later. Probably we will just run 5-10 tests for sanity reason

quarckster commented:
@oshtaier @mirzal it seems #187  does the trick, please check

mirzal commented:
I've been in touch with Walden from dev team and here's what I've learned:

* Following pages are written entirely in React: Red Hat Repositories, Subscriptions (+ subpages) and Select Organization
* The best way of ensuring that React has finished loading is hooking into the redux store and investigating redux states
    * But redux store is not exposed to external world. Technically it could, but such change is unlikely to get accepted by dev team.
* "Somewhat fragile" way is ensuring that no spinners exist - which is basically what I do with this PR
* There are additional React components used in Satellite ([full list](https://github.com/theforeman/foreman/tree/develop/webpack/assets/javascripts/react_app/components)). They probably require different handling than React pages, if any. I add this for reference only, as we might want to pay special attention to them.

I'm going to run selected tests with that change applied and will report results here once I have them.

mirzal commented:
@oshtaier , @abalakh please read comment below and let me know if that addresses your concerns. If everything is good, I would like to merge it.

## Impact on test results

I have run arbitrary subset of test with that change and without it. Results are below.

The bottom line is: tests that passed without this change also pass after applying it. Two tests failed without the patch but did pass with it; I think it has more to do with their flakiness than with proposed change.

Test id|Before change|After change
---|---|---
test_activationkey.py::test_positive_search_scoped|PASSED|PASSED
test_activationkey.py::test_positive_create_with_host_collection|PASSED|PASSED
test_activationkey.py::test_positive_create_with_envs|PASSED|PASSED
test_activationkey.py::test_positive_add_host_collection_non_admin|PASSED|PASSED
test_activationkey.py::test_positive_remove_host_collection_non_admin|PASSED|PASSED
test_activationkey.py::test_positive_delete_with_env|PASSED|PASSED
test_activationkey.py::test_positive_update_env|PASSED|PASSED
test_activationkey.py::test_positive_add_custom_product|PASSED|PASSED
test_activationkey.py::test_positive_fetch_product_content|PASSED|PASSED
test_contentcredential.py::test_positive_add_empty_product|FAILED|PASSED
test_contentcredential.py::test_positive_add_product_with_repo|PASSED|PASSED
test_contentcredential.py::test_positive_add_product_with_repos|PASSED|PASSED
test_contentcredential.py::test_positive_add_repo_from_product_with_repo|PASSED|PASSED
test_contentcredential.py::test_positive_add_repo_from_product_with_repos|PASSED|PASSED
test_contentcredential.py::test_positive_update_key_for_empty_product|PASSED|PASSED
test_contentcredential.py::test_positive_update_key_for_product_with_repo|PASSED|PASSED
test_contentcredential.py::test_positive_update_key_for_product_with_repos|PASSED|PASSED
test_contentcredential.py::test_positive_update_key_for_repo_from_product_with_repo|PASSED|PASSED
test_contentcredential.py::test_positive_update_key_for_repo_from_product_with_repos|PASSED|PASSED
test_contentcredential.py::test_positive_delete_key_for_empty_product|PASSED|PASSED
test_contentcredential.py::test_positive_delete_key_for_product_with_repo|PASSED|PASSED
test_contentcredential.py::test_positive_delete_key_for_product_with_repos|PASSED|PASSED
test_contentcredential.py::test_positive_delete_key_for_repo_from_product_with_repo|PASSED|PASSED
test_contentcredential.py::test_positive_delete_key_for_repo_from_product_with_repos|PASSED|PASSED
test_contentview.py::test_positive_add_puppet_module|FAILED|PASSED
test_contentview.py::test_negative_add_puppet_repo_to_composite|PASSED|PASSED
test_contentview.py::test_negative_add_components_to_non_composite|PASSED|PASSED
test_contentview.py::test_positive_add_non_composite_cv_to_composite|PASSED|PASSED
test_contentview.py::test_negative_add_dupe_modules|PASSED|PASSED
test_location.py::test_positive_update_domain|PASSED|PASSED
test_location.py::test_positive_update_user|PASSED|PASSED
test_location.py::test_positive_update_hostgroup|PASSED|PASSED
test_location.py::test_positive_add_org|PASSED|PASSED
test_location.py::test_update_environment|PASSED|PASSED
test_location.py::test_positive_update_medium|PASSED|PASSED
test_location.py::test_positive_update_template|PASSED|PASSED
test_organization.py::test_positive_update_user|PASSED|PASSED
test_organization.py::test_positive_update_media|PASSED|PASSED
test_organization.py::test_positive_update_template|PASSED|PASSED
test_organization.py::test_positive_update_ptable|PASSED|PASSED
test_organization.py::test_positive_update_domain|PASSED|PASSED
test_organization.py::test_positive_update_environment|PASSED|PASSED
test_organization.py::test_positive_update_hostgroup|PASSED|PASSED
test_organization.py::test_positive_update_location|PASSED|PASSED

## Impact on performance

I ran another subset of tests and measured how much time it took them to complete with patch and without it. I repeated that three times to get some idea about variance.

I believe that impact on performance is negligible. Some tests and some cases ran faster with patch, some ran slower, but variation is so big, that this is unlikely to matter in grand scheme of things. Some stats are below, I can share raw data if anyone thinks it's worth running more sophisticated analysis.

Global means and standard deviation for time (in seconds):

Group|Mean|Standard deviation
---|---|---
Before change| 89.1|30.0
After change|88.6|30.3

By case:

Case | Before change mean | Before change stddev| After change mean | After change stddev
---|---|---|---|---
Simple test (test_activationkey.py::test_positive_create)|55.6|13.4|38.2|5.95
Tier 2 test (test_contentview.py::test_positive_promote_with_custom_content)|141.|0.778|143.|21.9
Entire file (test_organization.py)|801.|134.|794.|30.2

And also by test, this time as chart. Bars around means represent one standard deviation:

![tests-plot](https://user-images.githubusercontent.com/12373754/45571121-7d685f80-b865-11e8-87a3-f3663c591269.png)
